### PR TITLE
Extend and remove recurring appointments

### PR DESCRIPTION
**Pull request by venkatavineela**

### CODE REVIEW

File: omod/src/main/java/org/openmrs/module/appointments/web/service/AbstractRecurringAppointmentsGenerationService.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
+package org.openmrs.module.appointments.web.service;
+import org.openmrs.api.APIException;
+import org.openmrs.module.appointments.model.Appointment;
+import org.openmrs.module.appointments.model.AppointmentRecurringPattern;
+import org.openmrs.module.appointments.model.AppointmentStatus;
+import org.openmrs.module.appointments.web.contract.AppointmentRequest;
+import java.util.*;
+import java.util.stream.Collectors;
+public abstract class AbstractRecurringAppointmentsGenerationService {
```
Comment (sowmika148) [inline omod/src/main/java/org/openmrs/module/appointments/web/service/AbstractRecurringAppointmentsGenerationService.java:9]: Rename class name

File: omod/src/main/java/org/openmrs/module/appointments/web/mapper/AllAppointmentRecurringPatternMapper.java [hunk 2]
```diff
@@ -60,17 +60,48 @@ public AppointmentRecurringPattern fromRequest(AppointmentRequest appointmentReq
...
         return updatedAppointmentRecurringPattern;
     }
+
     private List<Appointment> getUpdatedSetOfAppointments(AppointmentRequest appointmentRequest, AppointmentRecurringPattern appointmentRecurringPattern) {
         List<Appointment> newSetOfAppointments = new ArrayList<>();
         if (appointmentRecurringPattern.getEndDate() == null){
+            if (appointmentRequest.getRecurringPattern().getFrequency() > appointmentRecurringPattern.getFrequency()) {
+                newSetOfAppointments = addRecurringAppointments(appointmentRecurringPattern, appointmentRequest);
+            }
             if (appointmentRequest.getRecurringPattern().getFrequency() < appointmentRecurringPattern.getFrequency()) {
```
Comment (sowmika148) [inline omod/src/main/java/org/openmrs/module/appointments/web/mapper/AllAppointmentRecurringPatternMapper.java:70]: extract to method

File: omod/src/main/java/org/openmrs/module/appointments/web/service/impl/DailyRecurringAppointmentsGenerationService.java [hunk 3]
```diff
@@ -31,7 +31,7 @@ public DailyRecurringAppointmentsGenerationService(AppointmentRecurringPattern a
         return generateAppointments(endDate);
     }
 
-    private Date getEndDate() {
+    public Date getEndDate() {
```
Comment (sowmika148) [inline omod/src/main/java/org/openmrs/module/appointments/web/service/impl/DailyRecurringAppointmentsGenerationService.java:34]: private method

File: omod/src/main/java/org/openmrs/module/appointments/web/service/impl/WeeklyRecurringAppointmentsGenerationService.java [hunk 4]
```diff
@@ -34,35 +29,31 @@ public WeeklyRecurringAppointmentsGenerationService(AppointmentRecurringPattern
         return generateAppointments(endDate);
     }
 
-    private Date getEndDate() {
+    public Date getEndDate() {
```
Comment (sowmika148) [inline omod/src/main/java/org/openmrs/module/appointments/web/service/impl/WeeklyRecurringAppointmentsGenerationService.java:32]: private method

File: omod/src/main/java/org/openmrs/module/appointments/web/service/impl/WeeklyRecurringAppointmentsGenerationService.java [hunk 5]
```diff
@@ -91,14 +87,41 @@ private void getSortedSelectedDayCodes(int startDayCode, List<Integer> selectedD
         return appointments;
     }
 
+    @Override
+    public List<Appointment> addAppointments() {
```
Comment (sowmika148) [inline omod/src/main/java/org/openmrs/module/appointments/web/service/impl/WeeklyRecurringAppointmentsGenerationService.java:91]: see if can extract into methods

File: omod/src/main/java/org/openmrs/module/appointments/web/service/impl/WeeklyRecurringAppointmentsGenerationService.java [hunk 6]
```diff
@@ -119,6 +142,21 @@ private void updateCalendarInstancesToNextSelectedDay(Calendar startCalender, Ca
         endCalender.add(Calendar.DAY_OF_WEEK, daysToBeAdded);
     }
 
+    private List<Integer> getSelectedDayCodesForExtend(List<Integer> selectedDaysPerWeek, int dayCode) {
+        int dayCodeIndex = 0;
+        List<Integer> updatedSelectedDayCodes = new ArrayList<>();
+        for (int i = 0; i < selectedDaysPerWeek.size(); i++) {
```
Comment (sowmika148) [inline omod/src/main/java/org/openmrs/module/appointments/web/service/impl/WeeklyRecurringAppointmentsGenerationService.java:148]: variable name

File: omod/src/main/java/org/openmrs/module/appointments/web/service/impl/WeeklyRecurringAppointmentsGenerationService.java [hunk 7]
```diff
@@ -119,6 +142,21 @@ private void updateCalendarInstancesToNextSelectedDay(Calendar startCalender, Ca
...
+            if (dayCode == selectedDaysPerWeek.get(i)) {
+                dayCodeIndex = i;
+                break;
+            }
+        }
+        for (int i = dayCodeIndex + 1; i < selectedDaysPerWeek.size(); i++) {
```
Comment (sowmika148) [inline omod/src/main/java/org/openmrs/module/appointments/web/service/impl/WeeklyRecurringAppointmentsGenerationService.java:154]: variable name

File: omod/src/main/java/org/openmrs/module/appointments/web/service/AbstractRecurringAppointmentsGenerationService.java (hunk 1 above)
Comment (venkatavineela) [reply omod/src/main/java/org/openmrs/module/appointments/web/service/AbstractRecurringAppointmentsGenerationService.java:9]: Renamed

**Pull request merged**